automation/run/store.go

```diff
@@ -0,0 +1,306 @@
+	Bus() event.Bus
+}
+
+// MemStore is an in memory representation of a Store
+type MemStore struct {
+	mu              sync.Mutex
+	bus             event.Bus
+	workflows       map[workflow.ID]*workflowMeta
+	runs            map[string]*State
+	handleRunEvents event.Handler
+}
+
+type workflowMeta struct {
+	count  int
+	runIDs []string
+}
+
+func newWorkflowMeta() *workflowMeta {
+	return &workflowMeta{
+		count:  0,
+		runIDs: []string{},
+	}
+}
+
+var _ Store = (*MemStore)(nil)
+
+// NewMemStore returns a MemStore
+func NewMemStore(bus event.Bus) *MemStore {
+	s := &MemStore{
+		bus:       bus,
+		workflows: map[workflow.ID]*workflowMeta{},
+		runs:      map[string]*State{},
+	}
+	s.handleRunEvents = HandleRunEventsFactory(s)
+	return s
+
+}
+
+// Put adds a run.State to a MemStore
+func (s *MemStore) Put(r *State) (*State, error) {
+	if r == nil {
+		return nil, ErrNilRun
+	}
+	run := &State{}
+	run.Copy(r)
+	if run.ID != "" {
+		fetchedR, err := s.Get(run.ID)
+		if errors.Is(err, ErrNotFound) {
+			return nil, ErrNotFound
+		}
+		if fetchedR.WorkflowID != run.WorkflowID {
+			return nil, ErrPutWorkflowIDMismatch
+		}
+	}
+	if run.ID == "" {
+		run.ID = NewID()
+	}
+	if err := run.Validate(); err != nil {
+		return nil, err
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	_, ok := s.workflows[run.WorkflowID]
+	if !ok {
+		wf := newWorkflowMeta()
+		s.workflows[run.WorkflowID] = wf
+	}
+	s.runs[run.ID] = run
+	s.workflows[run.WorkflowID].count++
+	runIDs := s.workflows[run.WorkflowID].runIDs
+	s.workflows[run.WorkflowID].runIDs = append(runIDs, run.ID)
+	return run, nil
+}
+
+// Get fetches a run.State using the associated ID
+func (s *MemStore) Get(id string) (*State, error) {
+	s.mu.Lock()
+	r, ok := s.runs[id]
+	s.mu.Unlock()
+	if !ok {
+		return nil, ErrNotFound
+	}
+	return r, nil
+}
+
+// Count returns the number of runs for a given workflow.ID
+func (s *MemStore) Count(wid workflow.ID) (int, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	wfm, ok := s.workflows[wid]
+	if !ok {
+		return 0, fmt.Errorf("%w %q", ErrUnknownWorkflowID, wid)
+	}
+	return wfm.count, nil
+}
+
+// List lists all the runs associated with the workflow.ID in reverse
+// chronological order
+func (s *MemStore) List(wid workflow.ID, limit, offset int) ([]*State, error) {
+	fetchAll := false
+	switch {
+	case limit == -1 && offset == 0:
+		fetchAll = true
+	case limit < 0:
+		return nil, fmt.Errorf("limit of %d is out of bounds", limit)
+	case offset < 0:
+		return nil, fmt.Errorf("offset of %d is out of bounds", offset)
+	case limit == 0:
+		return []*State{}, nil
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	wfm, ok := s.workflows[wid]
+	if !ok {
+		return nil, fmt.Errorf("%w %q", ErrUnknownWorkflowID, wid)
+	}
+	runIDs := wfm.runIDs
+	runs := []*State{}
+	for i := len(runIDs) - 1; i >= 0; i-- {
+		id := runIDs[i]
+		run, ok := s.runs[id]
+		if !ok {
+			return nil, fmt.Errorf("run %q missing from the store", id)
+		}
+		runs = append(runs, run)
+	}
+
+	if offset > len(runs) {
```

Comment:
Use `offset >= len(runs)`, since indexing an array at an index equal to its length is invalid.

Reply:
thank youuuuuuuuu